boards: Add non-secure target for nrf54L15dk

Vge0rge opened the pull request:
This adds the nrf54l15dk/nrf54l15/cpuapp/ns board variant to
    Zephyr. It allows to build applications for this target.
    
    This is an initial support for the non secure target which allows
    building and running tfm_ipc and config_build.
    
    This is NOT full support of the non secure target in upstream
    Zephyr.
    
    There are important limitations, such as:
    - The hardware Crypto accelerator is not supported and thus the non
      secur target is NOT secure for production applicatiions in upstream
      Zephyr.
    - The BL2 is not supported, so no DFU is supported with this support
    
    Most of the code chagnes here are taken from nRF Connect SDK
    in order to avoid having noups there.
    
    Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
samples: tfm: Add nRF54L15 in TF-M samples yaml

Add nRF54L15 non secure target for the config_build and
the tfm_ipc samples.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
manifest: Bring TF-M with nRF54L15 upstream support

Update west

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

nordicjm commented on boards/nordic/nrf54l15dk/Kconfig:
```diff
@@ -0,0 +1,34 @@
+
+if BOARD_NRF54L15DK_NRF54L15_CPUAPP_NS
+
+config NRF_MPC_REGION_SIZE
+	hex
+	default 0x1000
+	help
+	  Region size for the Memory Protection Controller (MPC) in bytes.
```
move to dts (see nrf7x in ncs)

Vge0rge replied:
nrf7 doesn't support TF-M yet. I will have to check what sideffects this will give. I would prefer to keep the scope of this PR as is since its is already big and create a task to move to DTS later. Then the change can happen in NCS and in upstream Zephyr at the same time.

nordicjm replied:
this is a hardware property, it needs to move to dts, it's value can still be accessed by Kconfig using the devicetree functions in Kconfig. Move this binding upstream https://github.com/nrfconnect/sdk-nrf/blob/main/dts/bindings/mpc/nordic%2Cnrf-mpc.yaml then add it to dts with https://github.com/nrfconnect/sdk-nrf/blob/main/dts/common/nordic/nrf7120_enga.dtsi#L80

Vge0rge replied:
I added the nrf-mpc dts entry. Have a look again.

nordicjm commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns.dts:
```diff
@@ -0,0 +1,100 @@
+		 * header flash_layout.h of the relevant platform. Any updates in the layout
+		 * needs to happen both in the flash_layout.h and in this file at the same time.
+		 */
+		slot0_partition: partition@0 {
```
spacing between nodes

Vge0rge replied:
will update, thanks!

nordicjm replied:
needs newline on line 66

kartben commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns.yaml:
```diff
@@ -0,0 +1,22 @@
+arch: arm
+toolchain:
+  - gnuarmemb
+  - xtools
```
```suggestion
```

Toolchain not being used anymore

Vge0rge replied:
Will do thanks!

Vge0rge committed:
boards: Add non-secure target for nrf54L15dk

This adds the nrf54l15dk/nrf54l15/cpuapp/ns board variant to
Zephyr. It allows to build applications for this target.

This is an initial support for the non secure target which allows
building and running tfm_ipc and config_build.

This is NOT full support of the non secure target in upstream
Zephyr.

There are important limitations, such as:
- The hardware Crypto accelerator is not supported and thus the non
  secur target is NOT secure for production applicatiions in upstream
  Zephyr.
- The BL2 is not supported, so no DFU is supported with this support

Most of the code chagnes here are taken from nRF Connect SDK
in order to avoid having noups there.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

tomi-font commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns_defconfig:
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# Enable hardware stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
+
+# Enable TrustZone-M
```
Are those comments (`Enable X`) really needed?

Vge0rge replied:
I will remove the ones that don't have any additional info

tomi-font commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns_defconfig:
```diff
@@ -0,0 +1,44 @@
+# Disable TFM BL2 since it is not supported
+CONFIG_TFM_BL2=n
+
+# Support fo silence logging is not supported at the moment
```
```suggestion
# Support for silence logging is not supported at the moment
```

tomi-font commented on modules/trusted-firmware-m/nordic/include/RTE_Device.h:
```diff
@@ -44,6 +44,19 @@
 	) \
 }
 
+#elif DT_PINCTRL_HAS_NAME(DT_NODELABEL(uart30), default)
+#define RTE_USART30 1
```
```suggestion
#elif DT_PINCTRL_HAS_NAME(DT_NODELABEL(uart30), default)

#define RTE_USART30 1
```

tomi-font commented on modules/trusted-firmware-m/nordic/include/tfm_peripherals_config.h:
```diff
@@ -12,9 +12,15 @@ extern "C" {
 #define TFM_PERIPHERAL_UARTE1_SECURE 1
 #endif
 
+#endif
```
```suggestion
#endif /* SECURE_UART1 */
```

tomi-font commented on modules/trusted-firmware-m/nordic/include/tfm_peripherals_config.h:
```diff
@@ -29,7 +35,10 @@ extern "C" {
+#elif defined(NRF54L15_XXAA)
+    #include <tfm_peripherals_config_nrf54l15.h>
 #else
+
```
```suggestion
```

tomi-font commented on modules/trusted-firmware-m/nordic/nrf54l15_cpuapp/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2023, Nordic Semiconductor ASA.
```
2023 :face_holding_back_tears:

Vge0rge commented on boards/nordic/nrf54l15dk/Kconfig.defconfig:
```diff
@@ -9,3 +13,27 @@ config ROM_START_OFFSET
+	default y if BOARD_NRF54L15DK_NRF54L15_CPUAPP_NS
+
+# By default, if we build with TF-M, instruct build system to
+# flash the combined TF-M (Secure) & Zephyr (Non Secure) image
```
I committed that too fast actually. This was correct before I think so it is referring to a single combined image.

Vge0rge commented on boards/nordic/nrf54l15dk/Kconfig:
```diff
@@ -0,0 +1,34 @@
+# nRF54L15 DK board configuration
```
Yeah, will do

Vge0rge commented on boards/nordic/nrf54l15dk/Kconfig.defconfig:
```diff
@@ -9,3 +13,27 @@ config ROM_START_OFFSET
+
+# By default, if we build with TF-M, instruct build system to
+# flash the combined TF-M (Secure) & Zephyr (Non Secure) image
+config TFM_FLASH_MERGED_BINARY
+	default y
+	depends on BUILD_WITH_TFM
```
You are right, I will remove this

Vge0rge commented on boards/nordic/nrf54l15dk/board.cmake:
```diff
@@ -8,5 +8,13 @@ elseif(CONFIG_SOC_NRF54L05_CPUFLPR OR CONFIG_SOC_NRF54L10_CPUFLPR OR CONFIG_SOC_
   board_runner_args(jlink "--device=RISC-V" "--speed=4000" "-if SW" "--tool-opt=-jlinkscriptfile ${JLINKSCRIPTFILE}")
 endif()
 
+if(BOARD_NRF54L15DK_NRF54L15_CPUAPP_NS)
```
good point, I will add that

Vge0rge commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns.dts:
```diff
@@ -0,0 +1,105 @@
+	 * needs to happen both in the flash_layout.h and in this file at the same time.
+	 */
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
```
What do you mean? In general why one need to add this address-cells and size-cells?
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#address-cells-and-size-cells

Vge0rge replied:
I updated the PR

Vge0rge committed:
samples: tests: Exclude nRF54L15 NS for unsupported tests/samples

The ADC and watchdog samples are not currently supported for the
non secure target of the nRF54L15 DK.

Add exclusions for these until they are supported.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns.dts:
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
```
Done

nordicjm commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_ns.dts:
```diff
@@ -0,0 +1,113 @@
+			reg = <0x20014000 DT_SIZE_K(80)>;
+		};
+	};
+
```
```suggestion
```

Vge0rge committed:
dts: bindings: Add binding for Nordic MPC

Add binding for the Nordic Memory
Privilege Controller (MPC).

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
dts: nordic: Add NRF MPC node in dts

Add the Nordic Memory Privilege Controller (MPC)
in the dtsi of nrf54l15/10/05.

This sets the number of override regions and the
granularity of the regions.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

tomi-font commented on west.yml:
```diff
@@ -353,7 +353,7 @@ manifest:
         - tee
     - name: trusted-firmware-m
-      revision: f04edd12b5e5de2e2291f6e0a15efc03924dcbc3
+      revision: pull/122/head
```
```suggestion
      revision: 25dff7328c77b51e314ef9d4625044072040c96d
```

Vge0rge commented:
Rebased and updated the commit hash.

tomi-font commented:
Somehow the bot didn't remove the DNM label, so removed it manually.
